### ISSUE TITLE
Trainable variables

### ISSUE THREAD

**Issue by rdancer** [rdancer/Tetris-2023#4]
Replace the coefficients in the Reward.get_reward() function with TensorFlow trainable variables

```python
import tensorflow as tf

class Tetris:
    def __init__(self):
        # Initialize TensorFlow variables for the coefficients
        self.keep_height_low_coefficient = tf.Variable(10.0, name="keep_height_low_coefficient")
        self.fill_rows_evenly_coefficient = tf.Variable(1.0, name="fill_rows_evenly_coefficient")
        self.keep_board_empty_coefficient = tf.Variable(1.0, name="keep_board_empty_coefficient")
        self.increase_score_coefficient = tf.Variable(1.0, name="increase_score_coefficient")
        self.careful_when_score_high_coefficient = tf.Variable(1.0, name="careful_when_score_high_coefficient")
        self.complete_rows_coefficient = tf.Variable(100.0, name="complete_rows_coefficient")
        
    def calculate_reward(self, before_state, after_state):
        keep_height_low = (self.last_empty_row(after_state["board"]) - self.last_empty_row(before_state["board"])) * self.keep_height_low_coefficient
        fill_rows_evenly = (sum(self.row_fill_fractions(after_state["board"])) - sum(self.row_fill_fractions(before_state["board"]))) * self.fill_rows_evenly_coefficient
        keep_board_empty = (self.board_fill_fraction(after_state["board"]) - self.board_fill_fraction(before_state["board"])) * self.keep_board_empty_coefficient
        increase_score = (after_state["score"] - before_state["score"]) * self.increase_score_coefficient
        careful_when_score_high = (after_state["highScore"] - before_state["highScore"]) * self.careful_when_score_high_coefficient
        complete_rows = (self.num_completed_rows(after_state["board"]) ** 1.5) * self.complete_rows_coefficient

        total_reward = keep_height_low + fill_rows_evenly + keep_board_empty + increase_score + careful_when_score_high + complete_rows
        return total_reward

# Usage:
game = Tetris()
before_state = {"board": ..., "score": ..., "highScore": ...}
after_state = {"board": ..., "score": ..., "highScore": ...}
reward = game.calculate_reward(before_state, after_state)
```

```python
def update_coefficients(self, learning_rate, before_state, after_state):
    """Update the coefficients of the Tetris class using the Adam optimizer, based on the reward calculated from the before and after states of a single move. It calculates the loss as the negative mean of the reward and uses the gradients of the loss with respect to the coefficients to update the values of the coefficients. The method returns the final values of the coefficients after updating them. It is designed to be called once per Tetris move, with the appropriate learning rate and the before and after states of the game.
    """
    optimizer = tf.optimizers.Adam(learning_rate)
    
    with tf.GradientTape() as tape:
        reward = self.calculate_reward(before_state, after_state)
        loss = -tf.reduce_mean(reward)
    gradients = tape.gradient(loss, [self.keep_height_low_coefficient, self.fill_rows_evenly_coefficient, self.keep_board_empty_coefficient, 
                                    self.increase_score_coefficient, self.careful_when_score_high_coefficient, self.complete_rows_coefficient])
    optimizer.apply_gradients(zip(gradients, [self.keep_height_low_coefficient, self.fill_rows_evenly_coefficient, self.keep_board_empty_coefficient, 
                                             self.increase_score_coefficient, self.careful_when_score_high_coefficient, self.complete_rows_coefficient]))
    return self.keep_height_low_coefficient, self.fill_rows_evenly_coefficient, self.keep_board_empty_coefficient, self.increase_score_coefficient, self.careful_when_score_high_coefficient, self.complete_rows_coefficient
```
